hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -298,8 +295,56 @@ object LowerTableIR {
                     })
                   }
                 )
+              }
+            }
+
+        // TODO: This ignores nPartitions and bufferSize
+        case TableKeyByAndAggregate(child, expr, newKey, nPartitions, bufferSize) =>
+          val loweredChild = lower(child)
+          val newKeyType = newKey.typ.asInstanceOf[TStruct]
+          val oldRowType = child.typ.rowType
+          val filteredOldRowType = oldRowType.filter(field => !newKeyType.fieldNames.contains(field.name))._1
+          val oldRowFieldsNotInNewKey = filteredOldRowType.fieldNames
+          val shuffledRowType = newKeyType ++ filteredOldRowType
+
+          val withNewKeyFields = loweredChild.mapPartition { partition =>
+            mapIR(partition) { partitionElement =>
+              Let("row",
+                partitionElement,
+                bindIR(newKey) { newKeyRef =>
+                  val getKeyFields = newKeyType.fieldNames.map(fieldName => fieldName -> GetField(newKeyRef, fieldName)).toIndexedSeq
+                  val getOtherFields = oldRowFieldsNotInNewKey.map(fieldName => fieldName -> GetField(partitionElement, fieldName)).toIndexedSeq
+                  MakeStruct( getKeyFields ++ getOtherFields)
+                }
```

Comment:
I think the most direct way to write this is using `InsertFields`, which handles overwriting existing fields for you:
```
InsertFields(partitionElement, getKeyFields)
```